src/renderers/WebGLRenderer.js

```diff
@@ -1821,6 +1821,10 @@ THREE.WebGLRenderer = function ( parameters ) {
 
 			return a.object.renderOrder - b.object.renderOrder;
 
+		} else if ( !!a.material.program && !!b.material.program && a.material.program.id !== b.material.program.id ) {
```

Comment:
@mrdoob I also want the option to be able to do buffer uploads independent of the render. Either to do at end of frame rather than start; or on every other RAf to cap at 30fps with housekeeping; so can look at it then?

As initMaterials and shader compilation would fit into the housekeeping phase.